### PR TITLE
33 fen tests

### LINKED ISSUE #33: FEN Tests
Create unit tests to ensure proper parsing and serialization of FEN strings and `Chessboard`s

### PR DESCRIPTION
**Pull request by chanson02**
Please let me know if you can think of any other edge cases to test!

### CODE REVIEW

File: src/board.rs [hunk 1]
```diff
@@ -106,7 +108,7 @@ impl Chessboard {
     ///
     /// # Example
     ///
-    /// ```
+    /// ```ignore
```
Comment (chanson02) [inline src/board.rs:111]: These are what I'm talking about in #57

File: src/fen_util.rs [hunk 2]
```diff
@@ -50,7 +50,8 @@ pub fn place_pieces(chessboard: &mut Chessboard, fen_rows: &str) {
 /// A boolean indicating whether the FEN string is valid.
 #[allow(unused_variables)]
 pub fn valid_fen(fen: &str) -> bool {
-    let regex = Regex::new(r"^\s*^(((?:[rnbqkpRNBQKP1-8]+\/){7})[rnbqkpRNBQKP1-8]+)\s([b|w])\s([K|Q|k|q]{1,4})\s(-|[a-h][1-8])\s(\d+\s\d+)$").unwrap();
+    let regex = Regex::new(r"^\s*^(((?:[rnbqkpRNBQKP1-8]+\/){7})[rnbqkpRNBQKP1-8]+)\s([b|w])\s(-|[K|Q|k|q]{1,4})\s(-|[a-h][1-8])\s(\d+\s\d+)$").unwrap();
+
```
Comment (chanson02) [inline src/fen_util.rs:54]: Not sure if I included this in the commit message, but I changed this to allow `-` for one of the sections.

File: src/fen_util.rs [hunk 3]
```diff
@@ -91,6 +92,22 @@ pub fn valid_fen(fen: &str) -> bool {
...
+}
+
+// TODO: Move this somewhere it makes more sense
+fn unique_chars(s: &str) -> bool {
+    let mut chars = std::collections::HashSet::new();
+    for c in s.chars() {
+        if !chars.insert(c) {
+            return false;
+        }
+    }
```
Comment (chanson02) [inline src/fen_util.rs:110]: I couldn't figure out how to say that `KK` was an invalid castle right. Maybe we should panic here instead of returning false.

Comment (chanson02) [reply src/fen_util.rs:110]: Maybe we can make an issue out of that. Panic if unable to parse.
But I'm worried about how that may constrict us in the future.

File: src/board.rs [hunk 4]
```diff
@@ -510,3 +512,12 @@ impl Chessboard {
 //             break
 //     return min_eval
 // }
+
+/// I've never used this before, but the linter wants it
```
Comment (abrahimt) [inline src/board.rs:516]: What's a linter?

File: src/fen_util.rs (hunk 3 above)
Comment (abrahimt) [reply src/fen_util.rs:110]: I think returning false is fine for now. I also feel like it's time to make a util file; what do you think?

File: tests/fen_tests.rs [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+fn test_invalid_fen() {
+    // Only positions
+    assert!(!fen_util::valid_fen(
+        "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
+    ));
+    assert!(!fen_util::valid_fen("8/8/8/8/8/8/8/8"));
+    assert!(!fen_util::valid_fen("8/7k/8/8/8/8/8/8"));
+
+    // Missing parts
+    assert!(!fen_util::valid_fen("8/8/8/8/8/8/8/8 w - h8 0")); //move number
```
Comment (abrahimt) [inline tests/fen_tests.rs:77]: Now that we have a move piece function, we should probably add a move number attribute to Chessboard.

File: tests/fen_tests.rs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+    let passant = "h8";
+    fen_util::parse_en_passant(&mut cb, passant);
+    assert!(cb.en_passant == 64);
+
+    let passant = "-"; // set it back to empty
+    fen_util::parse_en_passant(&mut cb, passant);
+    assert!(cb.en_passant == 0);
+
+    // INVALID
```
Comment (abrahimt) [inline tests/fen_tests.rs:125]: What does this mean?

Comment (abrahimt) [reply tests/fen_tests.rs:125]: An invalid en passant?

File: tests/fen_tests.rs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    cb.white_king = 0b100000;
+    assert!(fen_util::get_fen_placement(&cb) == "8/8/8/8/8/8/8/PRNBQK2");
+
+    cb = Chessboard::new();
+    assert!(fen_util::get_fen_placement(&cb) == "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR");
+
+    cb = Chessboard::empty();
+    cb.black_pawns = 0b11111111_11111111_11111111_11111111_11111111_11111111_11111111_11111111;
+    //let s = fen_util::get_fen_placement(&cb);
+    //println!("FEN PLACEMENT {}!!!", s);
```
Comment (abrahimt) [inline tests/fen_tests.rs:281]: If we're not using this, let's delete it.

**Review by abrahimt — CHANGES_REQUESTED** [review]
Good job on these tests! We really needed them. I left a few comments I'd like you to take a look at before I approve this. 

- Abe

File: src/board.rs (hunk 4 above)
Comment (chanson02) [reply src/board.rs:516]: Clippy is the winter we're using
Linter** autocorrect since I'm on my phone

File: tests/fen_tests.rs (hunk 7 above)
Comment (chanson02) [reply tests/fen_tests.rs:281]: I left it in, in case I thought of some more unit tests

File: tests/fen_tests.rs (hunk 6 above)
Comment (chanson02) [reply tests/fen_tests.rs:125]: In the context of `test_parse_en_passant` the invalid test means trying to have it parse something invalid. Ex A9 or H0

File: src/fen_util.rs (hunk 3 above)
Comment (chanson02) [reply src/fen_util.rs:110]: I see a way to make this function a little shorter

Comment (chanson02) [reply src/fen_util.rs:110]: Yes let's make a util
- unique_chars
- Min
- Max
What else?

**Review by abrahimt — APPROVED** [review]
